Sources/GRPCCore/Metadata.swift

```diff
@@ -14,5 +14,329 @@
  * limitations under the License.
  */
 
-// FIXME: placeholder.
-public typealias Metadata = [String: String]
+/// A collection of metadata key-value pairs, found in RPC streams.
+/// A key can have multiple values associated to it.
+/// Values can be either strings or binary data, in the form of `Data`.
+/// - Note: Binary values must have keys ending in `-bin`, and this will be checked when adding pairs.
+public struct Metadata: Sendable, Hashable {
```

Comment:
Just a thought: we should add `CustomStringConvertible` conformance too (we'll need to do the same for `Value` as well.